File: app/src/main/java/com/steven/moviesearch/ResultItemAdapter.java
package com.steven.moviesearch;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.steven.moviesearch.models.ResultItem;

import java.util.List;

public class ResultItemAdapter extends RecyclerView.Adapter<ResultItemAdapter.ResultItemHolder> {

	private static final String TAG = ResultItemAdapter.class.getName();
	private List<ResultItem> resultItems;
	private OnItemClickListener onItemClickListener;
	Context context;

	public interface OnItemClickListener {

		void onItemClick(int position);
	}

	public ResultItemAdapter(Context context) {
		this.context = context;
	}

	@Override
	public ResultItemHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
		return new ResultItemHolder(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout
				.list_item, viewGroup, false));
	}

	@Override
	public void onBindViewHolder(ResultItemHolder resultItemHolder, int i) {
		String titleYear;
		ResultItem currentResultItem = resultItems.get(i);
		if (!TextUtils.isEmpty(currentResultItem.getReleaseDate())) {
			titleYear = currentResultItem.getTitle() + " (" + currentResultItem.getYear() + ")";
		} else {
			titleYear = currentResultItem.getTitle();
		}
		resultItemHolder.titleTextView.setText(titleYear);
		Glide.with(context)
				.load("https://image.tmdb.org/t/p/w500" + currentResultItem.getPosterPath())
				.apply(RequestOptions.centerCropTransform())
				.into(resultItemHolder.posterImageView);
	}

	@Override
	public int getItemCount() {
		return resultItems.size();
	}

	public void swapResults(List<ResultItem> resultItems) {
		if (resultItems != null) {
			this.resultItems = resultItems;
			notifyDataSetChanged();
		}
	}

	public ResultItem getItem(int position) {
		return resultItems.get(position);
	}

	public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
		this.onItemClickListener = onItemClickListener;
	}

	class ResultItemHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

		private TextView titleTextView;
		private ImageView posterImageView;

		public ResultItemHolder(View itemView) {
			super(itemView);
			titleTextView = (TextView) itemView.findViewById(R.id.textView_title_year);
			posterImageView = (ImageView) itemView.findViewById(R.id.imageView_poster);
			itemView.setOnClickListener(this);
		}

		@Override
		public void onClick(View v) {
			onItemClickListener.onItemClick(getAdapterPosition());
		}
	}

}

File: app/src/main/java/com/steven/moviesearch/ResultItemAdapter.kt
package com.steven.moviesearch

import android.content.Context
import android.support.v7.widget.RecyclerView.Adapter
import android.support.v7.widget.RecyclerView.ViewHolder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.steven.moviesearch.models.ResultItem
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.list_item.view.*

class ResultItemAdapter(private var context: Context) : Adapter<ResultItemAdapter.ResultItemHolder>() {
	private var resultItems: List<ResultItem>? = null
	var onItemClickListener: ((Int) -> Unit) = {}

	override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ResultItemHolder {
		return ResultItemHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout
				.list_item, viewGroup, false))
	}

	override fun onBindViewHolder(resultItemHolder: ResultItemHolder, i: Int) {
		val currentResultItem = resultItems?.get(i) ?: return
		val titleYear: String?
		if (!currentResultItem.year.isNullOrEmpty()) {
			titleYear = currentResultItem.title + " (" + currentResultItem.year + ")"
		} else {
			titleYear = currentResultItem.title
		}
		resultItemHolder.containerView.textView_title_year?.text = titleYear
		when (currentResultItem.mediaType) {
			"movie" -> resultItemHolder.containerView.textView_type?.text = "Movie"
			"tv" -> resultItemHolder.containerView.textView_type?.text = "TV Series"
		}


		currentResultItem.posterPath?.let { path: String ->
			Glide.with(context)
					.load("https://image.tmdb.org/t/p/w500$path")
					.apply(RequestOptions.centerCropTransform())
					.into(resultItemHolder.containerView.imageView_poster)
		} ?: resultItemHolder.containerView.imageView_poster.setImageDrawable(null)
	}

	override fun getItemCount(): Int {
		return resultItems?.size ?: 0
	}

	fun swapResults(resultItems: List<ResultItem>?) {
		if (resultItems != null) {
			this.resultItems = resultItems
			notifyDataSetChanged()
		}
	}

	fun getItem(position: Int): ResultItem? {
		return resultItems?.get(position)
	}

	inner class ResultItemHolder(override val containerView: View)
		: ViewHolder(containerView), LayoutContainer, View.OnClickListener {
		init {
			itemView.setOnClickListener(this)
		}

		override fun onClick(v: View) {
			onItemClickListener(adapterPosition)
		}
	}

	companion object {
		private val TAG = ResultItemAdapter::class.java.name
	}

}
